### ISSUE TITLE
[Image-Thumbnails] Race condition while generating thumbnail

### ISSUE THREAD

**Issue by podarcis** [pimcore/pimcore#4027]
# Bug Report

Generating thumbnails may lead to corrupt image when request in performed in parallel. If a thumbnail doesn't exist or needs to be re-generated, ```PublicServicesController::thumbnailAction``` asks for the thumbnail path in non-deferred mode and starts generating. As there is currently no locking mechanism, the process (generation) may run in parallel if clients requested the image at the same time, which could lead to very high load. To make things worse, the ```Image\Thumbnail\Processor::process()``` instructs the image adapter to overwrite an existing thumbnail, which can lead to a corrupted resulting image.

My guess is, that processes try to read the thumbnail while it is being written. 

### Expected behavior

Thumbnail generation should work even if two or more clients start a request simultaneously. Ideally, the generation process uses some kind of locking in order to prevent n-generations in parallel. At least the resulting image should always be correct.

### Actual behavior

Thumbnail has a good chance of being corrupted.

### Steps to reproduce

1. Install latest Pimcore demo.
2. Save&Publish an asset (in order to invalidate its cache; ideally a large asset) or just clear temporary files.
3. Run something like ```for i in {1..100}; do curl -v -o img$i.png http://pimcore.example.com/examples/south-africa/image-thumb__53__galleryLightbox/img_0322.png& done``` in order to start requests in parallel.
4. Check filesize and checksums ```md5sum img* && ls -l img*``` -> some files have random filesize and are corrupted.

### Workaround

The thumbnail is saved to a temporary file and is atomically renamed to the final filename.

```diff
diff --git a/models/Asset/Image/Thumbnail/Processor.php b/models/Asset/Image/Thumbnail/Processor.php
index cef03523d1..394f1377d7 100644
--- a/models/Asset/Image/Thumbnail/Processor.php
+++ b/models/Asset/Image/Thumbnail/Processor.php
@@ -348,7 +348,11 @@ class Processor
             $format = $image->getContentOptimizedFormat();
         }
 
-        $image->save($fsPath, $format, $config->getQuality());
+        // save the new thumbnail and perform atomic rename
+        $tmpFsPath = $fsPath . uniqid('.tmp-', true);
+        $image->save($tmpFsPath, $format, $config->getQuality());
+       @rename($tmpFsPath, $fsPath);
+
         $generated = true;
 
         if ($contentOptimizedFormat) {
```

### Possible solution

Some kind of locking?

**Comment by brusch** [comment]
I think locking would be too much overhead, in my opinion the atomic rename approach should be sufficient enough.

**Comment by podarcis** [comment]
@brusch Alright. Feel free to use above patch (w/ corrected whitespace). We've got this patch running in production for over a month now and haven't had a single problem since then.
